crates/uv-resolver/src/graph_ops.rs

```diff
@@ -1,81 +1,80 @@
 use pep508_rs::MarkerTree;
-use petgraph::algo::greedy_feedback_arc_set;
-use petgraph::visit::{EdgeRef, Topo};
-use petgraph::{Directed, Direction, Graph};
+use petgraph::graph::NodeIndex;
+use petgraph::visit::EdgeRef;
+use petgraph::{Direction, Graph};
+use rustc_hash::FxHashMap;
+use std::collections::hash_map::Entry;
 
-/// A trait for a graph node that can be annotated with a [`MarkerTree`].
-pub(crate) trait Markers {
-    fn set_markers(&mut self, markers: MarkerTree);
-}
-
-/// Propagate the [`MarkerTree`] qualifiers across the graph.
+/// Determine the markers under which a package is reachable in the dependency tree.
 ///
-/// The graph is directed, so if any edge contains a marker, we need to propagate it to all
-/// downstream nodes.
-pub(crate) fn propagate_markers<T: Markers>(
-    mut graph: Graph<T, MarkerTree, Directed>,
-) -> Graph<T, MarkerTree, Directed> {
-    // Remove any cycles. By absorption, it should be fine to ignore cycles.
-    //
-    // Imagine a graph: `A -> B -> C -> A`. Assume that `A` has weight `1`, `B` has weight `2`,
-    // and `C` has weight `3`. The weights are the marker trees.
-    //
-    // When propagating, we'd return to `A` when we hit the cycle, to create `1 or (1 and 2 and 3)`,
-    // which resolves to `1`.
-    //
-    // TODO(charlie): The above reasoning could be incorrect. Consider using a graph algorithm that
-    // can handle weight propagation with cycles.
-    let edges = {
-        let mut fas = greedy_feedback_arc_set(&graph)
-            .map(|edge| edge.id())
-            .collect::<Vec<_>>();
-        fas.sort_unstable();
-        let mut edges = Vec::with_capacity(fas.len());
-        for edge_id in fas.into_iter().rev() {
-            edges.push(graph.edge_endpoints(edge_id).unwrap());
-            graph.remove_edge(edge_id);
-        }
-        edges
-    };
-
-    let mut topo = Topo::new(&graph);
-    while let Some(index) = topo.next(&graph) {
-        let marker_tree = {
-            // Fold over the edges to combine the marker trees. If any edge is `None`, then
-            // the combined marker tree is `None`.
-            let mut edges = graph.edges_directed(index, Direction::Incoming);
+/// The algorithm is a variant of Dijkstra's algorithm for not totally ordered distances:
+/// Whenever we find a shorter distance to a node (a marker that is not a subset of the existing
+/// marker), we re-queue the node and update all its children. This implicitly handles cycles,
+/// whenever we re-reach a node through a cycle the marker we have is a more
+/// specific marker/longer path, so we don't update the node and don't re-queue it.
+pub(crate) fn marker_reachability<T>(
```

Comment:
This function only moved.

Reply:
If possible I might suggest splitting these moves out into separate commits in the future. Otherwise it can be a little tricky to see which parts of the diff are "uninteresting moves" versus what is an interesting change.

(It can be hard to always do this though, especially if you do the move mid-refactor.)

Reply:
Good call: #7091